src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Loader;
+
+use Symfony\Component\Config\Resource\DirectoryResource;
+
+/**
+ * DirectoryLoader is a recursive loader to go through directories
+ *
+ * @author Sebastien Lavoie <seb@wemakecustom.com>
+ */
+class DirectoryLoader extends FileLoader
+{
+    /**
+     * @param mixed  $file The resource
+     * @param string $type The resource type
+     */
+    public function load($file, $type = null)
+    {
+        $file = rtrim($file, '/');
+        $path = $this->locator->locate($file);
+        $this->container->addResource(new DirectoryResource($path));
+
+        foreach (scandir($path) as $dir) {
+            if ($dir[0] !== '.') {
+                if (is_dir("$path/$dir")) {
+                    $dir .= '/'; // append / to allow recursion
+                }
+
+                $this->setCurrentDir($path);
+
+                $this->import($dir, null, false, $path);
+            }
+        }
+    }
+
+    /**
+     * Returns true if this class supports the given resource.
+     *
+     * @param mixed  $resource A resource
+     * @param string $type     The resource type
+     *
+     * @return bool true if this class supports the given resource, false otherwise
+     */
+    public function supports($resource, $type = null)
+    {
+        return 'directory' === $type || (!$type && preg_match('/\/$/', $resource) === 1); // ends with a slash
```

Comment:
After extensive testing, I had to remove `is_dir` checks in both loaders because `->supports()` is not aware of the current location. Therefore, trailing slash is mandatory. I also updated the tests to reflect this.